juju/juju
# feature(secrets): do not update secret content if the same as before (WIP)

Updating a secret with the same content as is in the latest revision currently creates a new revision. There's been a request to make this idempotent.

The challenge is that for externally hosted secrets, the juju controller does not get to see the secret content. So the solution is to store the checksum of the content - the checksum of the most recent revision is now a secret metadata attribute. When new content is proposed to be saved, if its checksum matches that on the metadata then it becomes a no op.

For charm owned secrets, the metadata check is done in the unit agent as that's where the content is processed. For user secrets, the check is done in the facade for the same reason.

Care was needed to allow updates which included other attributes like labels etc to be saved even if the content itself was a no op.

For migration, juju/description was updated to handle the new checksum metadata attribute.

## QA steps

bootstrap and add a model and deploy a charm

```
$ juju bootstrap ...
$ juju add-model foo
$ juju deploy ubuntu
```

create same secret twice
check revision

```
$ juju exec -u ubuntu/0 -- secret-add foo=bar
secret://b55231c2-d9a5-494c-8cc3-59236e4b0204/cq5m4bcqqdhi05ss8c1g
$ juju exec -u ubuntu/0 -- secret-set cq5m4bcqqdhi05ss8c1g foo=bar
$ juju secrets
ID                    Name  Owner       Rotation  Revision  Last updated
cq5m4bcqqdhi05ss8c1g  -     controller  never            1  40 seconds ago  
$ juju exec -u ubuntu/0 -- secret-info-get cq5m4bcqqdhi05ss8c1g
cq5m4bcqqdhi05ss8c1g:
  revision: 1
  label: ""
  owner: application
  rotation: never
  ```

update just the label, no content

```  
$ juju exec -u ubuntu/0 -- secret-set cq5m4bcqqdhi05ss8c1g --label=foo
$ juju exec -u ubuntu/0 -- secret-info-get cq5m4bcqqdhi05ss8c1g
cq5m4bcqqdhi05ss8c1g:
  revision: 1
  label: foo
  owner: application
  rotation: never
 ```
 
update the content

```
$ juju exec -u ubuntu/0 -- secret-set cq5m4bcqqdhi05ss8c1g foo=bar2
$ juju exec -u ubuntu/0 -- secret-info-get cq5m4bcqqdhi05ss8c1g
cq5m4bcqqdhi05ss8c1g:
  revision: 2
  label: foo
  owner: application
  rotation: never
$ juju exec -u ubuntu/0 -- secret-get cq5m4bcqqdhi05ss8c1g
foo: bar2
$ juju show-secret cq5m4bcqqdhi05ss8c1g --reveal
cq5m4bcqqdhi05ss8c1g:
  revision: 2
  checksum: c839e1420b4f09f2952ab846c9741633c4b5d9965abe2a22e5f3fbefc0dfecf6
  owner: controller
  label: foo
  created: 2024-07-08T03:51:09Z
  updated: 2024-07-08T03:52:58Z
  content:
    foo: bar2
```

repeat for user secrets and secrets stored in external backend, eg vault

bootstrap another controller
migrate model foo to new controller
check that updating content to foo=bar2 keeps same revision

## Documentation changes

@tmihoc we'll need to make a small tweak to the secret docs

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2069238

**Jira card:** [JUJU-6225](https://warthogs.atlassian.net/browse/JUJU-6225)
